program_transformers/src/rollups/rollup_persister.rs

```diff
@@ -0,0 +1,571 @@
+            seq: value.seq,
+            index: value.index,
+        }
+    }
+}
+impl From<blockbuster::programs::bubblegum::ChangeLogEventV1> for ChangeLogEventV1 {
+    fn from(value: blockbuster::programs::bubblegum::ChangeLogEventV1) -> Self {
+        Self {
+            id: value.id,
+            path: value.path.into_iter().map(Into::into).collect::<Vec<_>>(),
+            seq: value.seq,
+            index: value.index,
+        }
+    }
+}
+
+impl From<&RolledMintInstruction> for BubblegumInstruction {
+    fn from(value: &RolledMintInstruction) -> Self {
+        let hash = value.leaf_update.hash();
+        Self {
+            instruction: InstructionName::MintV1,
+            tree_update: Some((&value.tree_update).into()),
+            leaf_update: Some(LeafSchemaEvent::new(
+                Version::V1,
+                value.leaf_update.clone(),
+                hash,
+            )),
+            payload: Some(Payload::MintV1 {
+                args: value.mint_args.clone(),
+                authority: value.authority,
+                tree_id: value.tree_update.id,
+            }),
+        }
+    }
+}
+
+#[automock]
+#[async_trait]
+pub trait RollupDownloader {
+    async fn download_rollup(&self, url: &str) -> Result<Box<Rollup>, RollupValidationError>;
+    async fn download_rollup_and_check_checksum(
+        &self,
+        url: &str,
+        checksum: &str,
+    ) -> Result<Box<Rollup>, RollupValidationError>;
+}
+
+pub struct RollupPersister<T: ConnectionTrait + TransactionTrait, D: RollupDownloader> {
+    txn: Arc<T>,
+    downloader: D,
+    cl_audits: bool,
+}
+
+pub struct RollupDownloaderForPersister {}
+
+#[async_trait]
+impl RollupDownloader for RollupDownloaderForPersister {
+    async fn download_rollup(&self, url: &str) -> Result<Box<Rollup>, RollupValidationError> {
+        let response = reqwest::get(url).await?.bytes().await?;
+        Ok(Box::new(serde_json::from_slice(&response)?))
+    }
+
+    async fn download_rollup_and_check_checksum(
+        &self,
+        url: &str,
+        checksum: &str,
+    ) -> Result<Box<Rollup>, RollupValidationError> {
+        let response = reqwest::get(url).await?.bytes().await?;
+        let file_hash = xxhash_rust::xxh3::xxh3_128(&response);
+        let hash_hex = hex::encode(file_hash.to_be_bytes());
+        if hash_hex != checksum {
+            return Err(RollupValidationError::InvalidDataHash(
+                checksum.to_string(),
+                hash_hex,
+            ));
+        }
+        Ok(Box::new(serde_json::from_slice(&response)?))
+    }
+}
+
+impl<T: ConnectionTrait + TransactionTrait, D: RollupDownloader> RollupPersister<T, D> {
+    pub fn new(txn: Arc<T>, downloader: D, cl_audits: bool) -> Self {
+        Self {
+            txn,
+            downloader,
+            cl_audits,
+        }
+    }
+
+    pub async fn persist_rollups(&self) {
+        loop {
+            tokio::time::sleep(Duration::from_secs(5)).await;
```

Comment:
Have you considered doing a a trigger/notification when a row is added to the `rollup_to_verify` table?  That would probably work better than a hardcoded delay for the times when there are no rollups and the times there are spikes of multiple new entries.

Reply:
Thank you! I added listener for inserts into `rollup_to_verify` table. Primary i was thinking about share single listener between multiple instances of RollupPersister but as far as most listener methods use mutable self reference (&mut self) i decided to create new listener for each RollupPersister. i hope that together with StartProcessing state it will not cause any race conditions

Reply:
I was also thinking about using single listener and sending notifications inside code using channels, but mpsc is very uncomfortable for such purpose because we need to send notifications for multiple consumers/workers. There also one community [crate](https://github.com/BurntSushi/chan) that implements Golang channels for Rust with multiple consumers but i think that it is bad practice to use additional import in our case, wdyt?

Reply:
I think there could be a race condition within `get_rollup_to_verify()`:

1. Both processes start a transaction, and both can use a `SELECT` query to find the same row in `rollups_to_verify`, which is a row that is not in the `StartProcessing` state.

2. Both processes attempt to update the selected row in `rollups_to_verify`, which does implicitly lock the row.  The second process waits for the lock on the row to be released.

3. The first process commits the transaction and releases the lock.  The second process successfully updates the row again.

4.  Now both processes move to `persist_rollup()` and can race each other in their respective state machines.

I think to solve this issue, there's a few options.

1. You could figure out the single listener thing and then each process would receive a notification.  However, if there's multiple rows getting added at nearly the same time, then there will be multiple triggers at nearly the same time.  From what I can tell there's no specific `rollup_to_verify` row associated with a trigger, so then the multiple processes could still find the same row, causing the same concurrency issue.

2. You could use a multi-consumer/multi-producer channel.   I agree that mpsc is incorrect for this purpose.  I agree the community crate is not right either, as the README says it is end-of-life and deprecated.  The Tokio tutorial at https://tokio.rs/tokio/tutorial/channels suggests to use `async-channel` crate for multi-consumer/multi-producer, so you could look into that.

3. My recommended option is this third one, which is modify `get_rollup_to_verify()` to do a `SELECT ... FOR UPDATE`.  I believe you can accomplish this by simply adding `lock_for_update()` to the `SELECT`:
```
let rollup_to_verify = rollup_to_verify::Entity::find()
    .filter(condition)
    .order_by_asc(rollup_to_verify::Column::CreatedAtSlot)
+   .lock_for_update() // *************************** Lock the row for the current transaction
    .one(&multi_txn)
    .await
    .map_err(|e| ProgramTransformerError::DatabaseError(e.to_string()))?;
```
Then I think `persist_rollup` also needs to be modified to NOT service rollups in the `ReceivedTransaction` state.  Otherwise I think it will still grab anything that was just ingested by the program transformer.

Reply:
Thank you very much!